### PR TITLE
[DeviceSanitizer] Memory overhead statistics

### PR DESCRIPTION
**Pull request by AllanZyne**
LLVM: https://github.com/intel/llvm/pull/14592

Statistic memory overhead (usm redzone + shadow memory) on each context, enabled by exporting asan flag `UR_LAYER_ASAN_OPTIONS=print_stats:1`.

Addtionally, move "AsanOptions" under "SanitizerInterceptor".

### CODE REVIEW

File: source/loader/layers/sanitizer/asan_statistics.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/asan_statistics.cpp:19]: Please don't use globals, they interfere with UR loader's ability to properly initialize and teardown.
Can't these variables be a part of the `AsanStats` struct?

Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_statistics.cpp:19]: Sorry, these variables are unused, I'll remove them.

Comment (pbalcer) [reply source/loader/layers/sanitizer/asan_statistics.cpp]: thanks

File: source/loader/layers/sanitizer/asan_statistics.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_statistics.cpp:22]: When print, can we also print the detailed data instead of only printing a simple overhead factor?

File: source/loader/layers/sanitizer/asan_interceptor.hpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_interceptor.hpp:119]: Why do we retain context but do not release them?

File: source/loader/layers/sanitizer/asan_options.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_options.cpp:101]: When AsanOptions is constructing, the getContext will refer to a also-constructing Context class. I think that is bad.

File: source/loader/layers/sanitizer/asan_statistics.cpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_statistics.cpp:22]: I didn't save detailed data. In the end, all saved data should be 0 if no memory leak.

File: source/loader/layers/sanitizer/asan_interceptor.hpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_interceptor.hpp:119]: It released here:

https://github.com/oneapi-src/unified-runtime/pull/1869/files#diff-cdb245c314b0deaa1e8277e335e881a8f7985456bb01ebcfe92aa4321b1de6ceR184-R186

File: source/loader/layers/sanitizer/asan_options.cpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_options.cpp:101]: offline sync with yingcong, we needn't to worry this.

**Comment by yingcong-wu** [comment]
It occurs to me that this only measures the redzone overhead, I think we should also mean the total overhead including shadow memory and other usages. Maybe we can implement a thin layer that keeps track of all memory allocation and compares the data with and without the sanitizer layer.

**Comment by AllanZyne** [comment]
> It occurs to me that this only measures the redzone overhead, I think we should also mean the total overhead including shadow memory and other usages. 

No, actually it includes shadow memory.

**Comment by AllanZyne** [comment]
Hi @pbalcer @kswiecicki, please review. Thanks!

File: source/loader/layers/sanitizer/asan_options.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/asan_options.cpp:23]: this will conflict with: https://github.com/oneapi-src/unified-runtime/pull/2054/files that I just merged. Please rebase.

**Review by pbalcer — APPROVED** [review]

File: source/loader/layers/sanitizer/asan_options.cpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_options.cpp]: Done

**Comment by AllanZyne** [comment]
Hi @pbalcer, I think the failed CIs are unrelated, could you please double check this? Thanks!

**Comment by pbalcer** [comment]
@AllanZyne they are indeed. I'll mark it ready to merge after you get approval from @yingcong-wu.

File: source/loader/layers/sanitizer/asan_statistics.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_statistics.cpp:1]: Just a idea and a nit, can we create a warpper to check `PrintStats`, and avoid (as much as we can) the overhead of this statistics when we not enable the tracking?

**Review by yingcong-wu — APPROVED** [review]

**Comment by pbalcer** [comment]
I resolved a conflict and rebased the PR.

**Comment by AllanZyne** [comment]
> I resolved a conflict and rebased the PR.
> 
> ... but it didn't work :-) please rebase and update the PR. Thanks.

No problem. Done.

File: source/loader/layers/sanitizer/asan_statistics.cpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_statistics.cpp:1]: Good idea. Done.

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Comment by AllanZyne** [comment]
Kindly ping @kswiecicki

**Pull request merged**
